### ISSUE TITLE
Report adverse events to the patient profile page and track it using the Immunization resource as a reaction

### ISSUE THREAD

**Issue by f-odhiambo** [opensrp/fhircore#227]
Once a patient has received the 1st dose of the vaccine, we can start tracking any reactions/symptoms that they exhibit post getting either the first shot or the second shot for a dual dose vaccine. The report adverse event will be recorded/updated onto the [Immunization resource](https://www.hl7.org/fhir/immunization.html) as a reaction i.e. Details of a reaction that follows the immunization. 

1. A list of adverse effects can also be found [here](https://www.cdc.gov/coronavirus/2019-ncov/vaccines/safety/adverse-events.html) 
2. What to Do if You Have an [Allergic Reaction After Getting A COVID-19 Vaccine](https://www.cdc.gov/coronavirus/2019-ncov/vaccines/safety/allergic-reaction.html)
3. [SNOMED CT Browser ](https://snomedbrowser.com/Codes/ConceptList?term=Thrombocytopenia&type=)for codable concepts/ condition  search 

The reaction field has a cardinality of 0..*

This will be captured using the following fields 
1. Date - When reaction started as a DateTime field
2. Detail  
3. Reported - Boolean  - Indicates self-reported reaction

**Acceptance criteria** 
- [ ] Show Report Adverse event button only after receiving the first shot 
- [ ] This will be persisted post the first step above
- [ ] Add a record adverse reaction page
- [ ] Save reaction to FHIR resources 
- [ ] List adverse reaction  page 

**Add Reaction** 
<img src="https://user-images.githubusercontent.com/12672919/134119760-421ac6b4-292c-47b1-8278-983edff265fd.jpg" width="200" height="400" />

**View Reactions**
<img src="https://user-images.githubusercontent.com/4540684/136219392-c2d490fb-e45c-4d7a-9c4b-cfb55ef014cc.png" width="200" height="400" />

**Comment by f-odhiambo** [comment]
@rowo We can have a follow up discussion about the design to support this implementation

**Comment by shoaibmushtaq25** [comment]
There are few questions after reading and digesting it. @f-odhiambo 
1. Is this the final design (attached screenshot) for this issue or do we still need to change it?
2. Where is the design for the item "Add a record adverse reaction page" in the acceptance criteria mentioned above?

**Comment by shoaibmushtaq25** [comment]
Right now, Exploring Immunization resource and how to use add adverse effects as reaction of Immunization resource.

**Comment by shoaibmushtaq25** [comment]
**Update**
Explored Immunization resource to use it to track reverse effects of vaccinations.
Now, waiting for respective designs to be ready. Meanwhile, I will be syncing up with @ekigamba around progress of extraction of multiple resources as we have multiple resources scenario in this Immunization case as well.

**Comment by rowo** [comment]
June 21 discussion with Francis

1. Does adverse reaction need to be associated with specific dose? No. 
2. Is adverse reaction form specific to a vaccine (J&J vs Pfizer)? No. 
3. One immunization/dose can have multiple reactions.
4. Does adverse reaction show forever? Yes. 
5. Do we need to display if someone has had an adverse reaction? Yes and read only. Doesn't impact ability to do second dose.
6. What's on the form? 1. Date. 2. Detail (select single and an other option)  3. Reported by (self or nurse).

two additional images needed from @rowo: form and way to see adverse reaction on patient.

**Comment by shoaibmushtaq25** [comment]
Here are the design for this feature https://www.figma.com/file/Ck46ofqC6kE6ISBC2YmwiI/COVID19-FHIR?node-id=1568%3A0

**Comment by shoaibmushtaq25** [comment]
Hey @f-odhiambo , Do we save these adverse effects of vaccination against a patient only on local database/ preferences or do we need to send it to server as well?

**Comment by f-odhiambo** [comment]
> Hey @f-odhiambo , Do we save these adverse effects of vaccination against a patient only on local database/ preferences or do we need to send it to server as well?

We need to update the Immunization resource being used to track the Vaccination . Will add the foreign keys of the reactions saved as  corresponding observations .

The Immunization resource will then be synced to HAPI to update the server copy

**Comment by shoaibmushtaq25** [comment]
As per the conversation with @f-odhiambo , we can use [Adverse Event](https://www.hl7.org/fhir/adverseevent.html#AdverseEvent) for the purpose of tracking Adverse Events of Vaccines and link them to the Patient resource.
We can track the adverse event details as follows,

1. Date - When the event occured as a DateTime field
2. Resulting Condition - Which is Reference to [Condition](https://www.hl7.org/fhir/condition.html)
3. We can track the conditions as code field (which is [CodeableConcept](https://www.hl7.org/fhir/datatypes.html#CodeableConcept)) of Condition. @f-odhiambo to add these codes for different conditions.

**Comment by f-odhiambo** [comment]
Based  on the Selected Adverse Events Reported after COVID-19 Vaccination listed [here](https://www.cdc.gov/coronavirus/2019-ncov/vaccines/safety/adverse-events.html) I have have found out the following codable code for the adverse events reported : 

**Codable concepts for the listed conditions** 
1.  Anaphylaxis - Concept ID: [39579001](https://snomedbrowser.com/Codes/Details/39579001)
2. Blood clots - Concept ID: [75753009](https://snomedbrowser.com/Codes/Details/75753009)
3. Myocarditis/Pericarditis  - Not same but different conditions 
4. Myocarditis - Concept ID: [50920009](https://snomedbrowser.com/Codes/Details/50920009)
5. Pericarditis  - Concept ID: [3238004](https://snomedbrowser.com/Codes/Details/3238004)
6. Thrombosis with thrombocytopenia does not exist in SNOMED DB
7. Thrombosis - Concept ID: [264579008](https://snomedbrowser.com/Codes/Details/264579008) 
8. Heparin-induced thrombocytopenia (disorder) Thrombosis (disorder) - Concept ID: [111588002](https://snomedbrowser.com/Codes/Details/111588002)


**_Let's use the below as an official list_**
1.  Anaphylaxis 
2. Blood clots 
3. Myocarditis 
4. Pericarditis  
8. Heparin-induced thrombocytopenia (disorder) Thrombosis (disorder) 

For the category  Other  - Not sure how will do this? Any thoughts @shoaibmushtaq25

**Comment by shoaibmushtaq25** [comment]
I am not sure about Category Other. Is it safe to use [note](https://www.hl7.org/fhir/condition-definitions.html#Condition.note) under [Condition](https://www.hl7.org/fhir/condition.html) for other category? WDYT @f-odhiambo

**Comment by f-odhiambo** [comment]
Sure that's sounds good otherwise other could make things quite complicated pretty fast

**Comment by shoaibmushtaq25** [comment]
**Update**
Had a call with Jing(Google), @ekigamba and Kashyap (IPRD) around the implementation of Adverse Events. 
There were two approaches which was under discussion.
1. [Immunization](https://www.hl7.org/fhir/immunization.html) with [reaction](https://www.hl7.org/fhir/immunization-definitions.html#Immunization.reaction) 
2. [AdverseEvent](https://www.hl7.org/fhir/adverseevent.html#AdverseEvent) with [resultingCondition](https://www.hl7.org/fhir/adverseevent-definitions.html#AdverseEvent.resultingCondition)

So out of these two approaches, we were agreed to go with the approach 1 because for the approach 2, [AdverseEvent](https://www.hl7.org/fhir/adverseevent.html#AdverseEvent) has the [MaturityLevel](https://www.hl7.org/fhir/versions.html#maturity) of 0 so it is not safe to use AdverseEvent right now.

After we have a consensus on using approach 1 (Immunization with reaction), we analyzed it in the way that whether we use [Defitinition based extraction](http://hl7.org/fhir/uv/sdc/2019May/extraction.html#definition-based-extraction) or [Structure map based extraction](http://hl7.org/fhir/uv/sdc/2019May/extraction.html#structuremap-based-extraction) to save the data. 
We found that, it looks difficult or even impossible to use Definition based extraction for this use case and hence we are going about using Structure map based extraction for this use case.
Also @ekigamba discussed it with @rkodev , he also agreed upon using the structure map based extraction for it.
So I am going with structure map based extraction approach.

@ekigamba Please review it and add/modify anything If I missed it.
cc: @f-odhiambo @pld

**Comment by shoaibmushtaq25** [comment]
**Update**
Working on updating JSON based on the Observation Reference ([detail](https://www.hl7.org/fhir/immunization-definitions.html#Immunization.reaction.detail) of [reaction](https://www.hl7.org/fhir/immunization-definitions.html#Immunization.reaction) under [Immunization](https://www.hl7.org/fhir/immunization.html)). Also extracting these questionnaire responses into Immunization resource and update and save the resource into FHIR engine in order to load it from FHIR engine afterwards when to show adverse events data into list view.

Learning about [Fhir-Mapping language](https://www.hl7.org/fhir/mapping-language.html) in order to make a structure map for Immunization with reaction ( for adverse events).

**Comment by shoaibmushtaq25** [comment]
Hey @ekigamba , Can you please review this Structure map for adverse reaction ? I have prepared it. Also please suggest me the way that how can I test that my structure map is correct or not? Thanks in advance.

`  map "http://hl7.org/fhir/StructureMap/AdverseReaction" = 'Adverse Reaction of vaccine'
  
  uses "http://hl7.org/fhir/StructureDefinition/QuestionnaireReponse" as source
  uses "http://hl7.org/fhir/StructureDefinition/Bundle" as target
  uses "http://hl7.org/fhir/StructureDefinition/Observation as source
  uses "http://hl7.org/fhir/StructureDefinition/Immunization as target
  
  group AdverseReaction(source src : QuestionnaireResponse, target bundle: Bundle) {
      src -> bundle.id = uuid() "rule_c";
      src -> bundle.type = 'collection' "rule_b";
      src -> bundle.entry as entry, entry.resource = create('Immunization') as immunization then
          ExtractImmunization(src, bundle, immunization) "rule_i";
  }
  
  group ExtractImmunization(source src: QuestionnaireResponse, target bundle: Bundle, target immunization: Immunization) {
      src -> immunization.id = uuid() "rule_j";
      
      src.item as item where(linkId = 'adverse-event-reaction') -> immunization.reaction = create('BackboneElement') as immunizationReaction then {
      
        item.item as inner_item where (linkId = 'adverse-event-date') then {
          inner_item.answer first as ans then {
            ans.value as val -> immunizationReaction.date = val "rule_a";
          };
        };
        
        item.item as reaction_detail_item where (linkId = 'adverse-event-codes') -> bundle.entry as entry, entry.resource = create('Observation') as observation then {
          reaction_detail_item -> observation.id = uuid() "rule_obs_1";
          reaction_detail_item -> observation.code = evaluate(reaction_detail_item, ${"$"}this.item.answer.value) "rule_obs_2";
        };

        observation -> immunizationReaction.detail = create('Reference') as observationReference then {
         observation.id as theObservationId -> observationReference.reference = theObservationId "rule_obs_3";
         observation -> observationReference.type = "Observation" "rule_obs_4";
         } "rule_obs_5";
      }
  }
  
`

**Comment by ekigamba** [comment]
@shoaibmushtaq25 
- We are editing a previous Immunization so the `Immunization.id` should be prepopulated on the QuestionnaireResponse and retrieved from the hidden field
- A few of the lines inside the `ExtractImmunization` group might fail. You'll need to validate this StructureMap
- The StructureDefinition going into `Immunization.reaction` is probably `Immunization.Immunzation_Reaction` as declared in the docs https://www.hl7.org/fhir/immunization-definitions.html#Immunization.reaction. Kindly check the StructureDefinition for Immunization

**Comment by shoaibmushtaq25** [comment]
Hey @f-odhiambo  cc: @ekigamba ,  Do we have flexibility to use DateTime in place of only Date on the screen below because we have [date](https://www.hl7.org/fhir/immunization-definitions.html#Immunization.reaction.date) as dateTime type under [reaction](https://www.hl7.org/fhir/immunization-definitions.html#Immunization.reaction) of [Immunization](https://www.hl7.org/fhir/immunization.html) resource? Actually, We are having some issue converting dateType to dateTimeType in structure map based extraction using FHIRPath because FHIRPath doesn't support this conversion so far and they still need to implement the utility method for this conversion.
![Screenshot from 2021-09-07 15-39-34](https://user-images.githubusercontent.com/12672919/132331837-31c79e2b-ec33-4b1d-915f-e6da71717a85.png)

**Comment by f-odhiambo** [comment]
@shoaibmushtaq25 We can change to DateTime format. Proceed

**Comment by pld** [comment]
Do we want users to select a time, or do we keep the UI the same so they're only entering a date but we store it as a date time?

**Comment by shoaibmushtaq25** [comment]
> Do we want users to select a time, or do we keep the UI the same so they're only entering a date but we store it as a date time?
> […](#)
> On Sep 8, 2021, at 03:07, Francis Odhiambo Otieno ***@***.***> wrote: ﻿ @shoaibmushtaq25 We can change to DateTime format. Proceed — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub, or unsubscribe. Triage notifications on the go with GitHub Mobile for iOS or Android.

I made the ui change, time to be entered by the users as shown below,
![Screenshot_20210908-145833_FHIR EIR](https://user-images.githubusercontent.com/12672919/132490208-104688ba-284f-4c54-b073-038b5c9aef22.jpg)

**Comment by pld** [comment]
This this make sense from a user perspective? Will they know the time that the adverse event took place?

**Comment by shoaibmushtaq25** [comment]
@pld Yeah, It's difficult to remember the exact time of the reaction. 
In that case, we have two options 

1. first one is showing both date and time on UI and user can enter the date and some random time of that date
2. second one could be showing the same UI as showing before (user have to enter only the date) and we should store it as DateTime instead (as you mentioned this above)

If we wanna proceed with option 2, I may need some help on that in order to make this conversion. We are using [this questionnaire](https://fhir.labs.smartregister.org/fhir/Questionnaire/1568) for the adverse events.

**QuestionnaireResponse** is here, 
`{"resourceType":"QuestionnaireResponse","item":[{"linkId":"adverse-event-reaction","item":[{"linkId":"adverse-event-codes","answer":[{"valueCoding":{"system":"https://www.snomed.org","code":"39579001","display":"Anaphylaxis"}}]},{"linkId":"adverse-event-date","answer":[{"valueDateTime":"2021-09-08T14:47:00+05:00"}]}]}]}`

and my **structure map for adverse events** is below,
`
    map "http://hl7.org/fhir/StructureMap/AdverseReaction" = 'AdverseReaction'

      uses "http://hl7.org/fhir/StructureDefinition/QuestionnaireReponse" as source
      uses "http://hl7.org/fhir/StructureDefinition/Bundle" as target
      uses "http://hl7.org/fhir/StructureDefinition/Observation" as source
      uses "http://hl7.org/fhir/StructureDefinition/Immunization" as target

      group AdverseReaction(source src : QuestionnaireResponse, target bundle: Bundle) {
          src -> bundle.id = uuid() "rule_c";
          src -> bundle.type = 'collection' "rule_b";
          src -> bundle.entry as entry, entry.resource = create('Immunization') as immunization then
              ExtractImmunization(src, bundle, immunization) "rule_i";
      }

      group ExtractImmunization(source src: QuestionnaireResponse, target bundle: Bundle, target immunization: Immunization) {
          src.id as srcId -> immunization.id = srcId "rule_j";

              src.item as item where(linkId = 'adverse-event-reaction') -> immunization.reaction = create('Immunization_Reaction') as immunizationReaction then {
                item.item as inner_item where (linkId = 'adverse-event-date') then {
              inner_item.answer first as ans then {
                ans.value as val -> immunizationReaction.date = val "rule_a";
              };
            };
          
            
            item.item as reaction_detail_item where (linkId = 'adverse-event-codes') -> bundle.entry as entry, entry.resource = create('Observation') as observation then {
              reaction_detail_item -> observation.id = uuid() "rule_obs_1";
              reaction_detail_item.answer as reactionAns -> observation.code = create('CodeableConcept') as codeableConcept then {
                reactionAns.value as reactionAnsValue -> codeableConcept.coding = reactionAnsValue "rule_obs_2_2";
              } "rule_ob_2";

             src -> immunizationReaction.detail = reference(observation) "rule_obs_3";
            };
          };
      };
  
  `

**Comment by pld** [comment]
OK I see why this is complex, I think for now let's leave it as a DateTime and come back to it if we get push back from the users or partners on having to enter the time.

**Comment by shoaibmushtaq25** [comment]
Just to clear my understanding on one thing, @f-odhiambo @pld @ekigamba 
We have 2 cases to report adverse events
1. Patient who has received only first dose of vaccine (Partially vaccinated), we receive a list of immunization of size 1 in that case.
2. Patient who has received two doses of vaccine (Fully vaccinated), we receive a list of immunization of size 2 in that case.

We show Report Adverse Event  button after receiving the first dose of vaccine, 
- For the 1st case, We will be editing the single immunization resource (that we are receiving against the patient)

**Question:** What about the 2nd case , we are receiving two immunization resources in that, which one to edit for the adverse events?

**Note:** We are tracking the adverse events as [detail](https://www.hl7.org/fhir/immunization-definitions.html#Immunization.reaction.detail) of  [reaction](https://www.hl7.org/fhir/immunization-definitions.html#Immunization.reaction) under the [immunization](https://www.hl7.org/fhir/immunization.html) resource.

**Comment by pld** [comment]
it'd make sense to me to edit the 2nd (later dated) immunization resources, if we're assuming that the adverse event is tied to the most recent immunization given, which seems reasonable

**Comment by f-odhiambo** [comment]
I think each adverse effect should be tagged to the immunisation to which you received and reacted. In the case of s single shot, a dual dose or a series of vaccinations, this may also apply. Not sure if this is what @pld  is alluding to but tagging it to the most recent vaccine may be erroneous. if indeed you are reacting to dose 1 and dose 2/n

@shoaibmushtaq25 does this create complexity in querying and displaying a list of reactions i.e The reaction and which immunization you reacted to?

**Comment by pld** [comment]
Cool, yeah I agree with Francis, that's what I was (convolutedly) alluding to.

If the assumption that the adverse event is related to the last received dose is incorrect, then I think we'd need a UI element to select the specific dose to attach the adverse event to

**Comment by shoaibmushtaq25** [comment]
Yeah, that's a good idea @pld to have a UI element for the selection of specific dose because adverse events would be attached to a specific dose. 
One more thing to consider here that the patient/specific user can report multiple adverse events as [reaction](https://www.hl7.org/fhir/immunization-definitions.html#Immunization.reaction) against each immunization/dose because cardinality of reaction is 0..*.  

In the other case, If we doesn't introduce a ui element for the selection, then we will be querying all the attached immunization resources to check the adverse events and display them in the list.

**Comment by shoaibmushtaq25** [comment]
@f-odhiambo , Can you please advise that what should we do in the case of **Others**? 
when a user doesn't select a adverse reaction from a pre-populated list given below but he selects **Others**
1. Anaphylaxis
2. Blood clots
3. Myocarditis
4. Pericarditis
5. Heparin-induced thrombocytopenia (disorder) Thrombosis (disorder)

Should we skip others for now or whats the better approach to cater this?
![adverse reaction with date time](https://user-images.githubusercontent.com/12672919/134119760-421ac6b4-292c-47b1-8278-983edff265fd.jpg)

**Comment by f-odhiambo** [comment]
I think we can skip the other category for now till we see how other IGs are using it. We can open another ticket to track this 
CC @shoaibmushtaq25

**Comment by f-odhiambo** [comment]
@RaaziaTarique Updated the main issue description 
CC @shoaibmushtaq25

**Comment by f-odhiambo** [comment]
@RaaziaTarique @shoaibmushtaq25 Any update on this?
